Integration rewrite and QoL improvements

karelkryda opened the pull request:
Hi @Pigotka,

It's time to submit a PR for this integration as well. As you can see, it is quite a lot of work. I will try to describe and explain everything.

## Improvements
* Added option to specify whether to bypass section errors by default
* Added option to specify how often to poll data from Jablotron Cloud
* Added option to specify time limit for polling data from Jablotron Cloud
* Integration (re)configuration dialog now remembers user input
* Added option to change integration settings - reconfigure
* Added automatic notification to change login credentials if they are no longer valid
* Entities now have the correct state right at creation
* Incorrect pin now returns a user friendly error
* Added information about the alarm control panel firmware
* Improved translations - NL is not complete

## Fixes
* Programmable gate control actions now correctly uses the default pin - fixes #47

## Changes
* Each control action and data polling uses a new API login - this should eliminate expired sessions
* Timeout when polling data throws a warning in the log instead of raising an error - entities are no longer marked as unavailable
* When performing control actions, the state of the entity is modified only in case of successful control
* Removed space from entity unique id attribute **- see notes**

## Testing
I tested the integration thoroughly and tried to simulate all scenarios that may occur.
Unfortunately I have neither temperature sensors nor programmable gates in my system. So I tested these entities by mocking the API data and everything seemed to work.
I didn't even notice any of the sections being marked as unavailable during testing.
I currently have a test instance of HA set up and am letting the integration run.

Please, also test the integration and verify that I didn't make a mistake somewhere during the rewrite before merging.

## Notes
**The unique entity ID contained a space, which didn't seem to be according to the rules. I fixed that, and added a method that migrates existing unique IDs with a space to new unique IDs with an underscore. This means that all entities should be preserved, just have their unique IDs fixed in the background. Users should not notice anything.**

---

It is now checked if the control action of the alarm and PGs was successful. If so, the corresponding entity state is set.
For alarms we could theoretically set the state to DISARMED/ARMED_HOME/ARMED_AWAY immediately, but for now I left it at ARMING/DISARMING.

---

In case an alarm is raised in a section, this information is not reflected by changing the state of the section in API response, but is written to the list of current events. Unfortunately, the information about which section the alarm is related to is not quite appropriately represented, so I did not include this functionality - i.e. reactions to alarms - in this PR.
We can talk about this separately.

---

The integration does not automatically add or remove entities. To find new devices and remove old ones, you must reinitialize the integration.
This behavior is the same as in the current version.

Koky05 commented:
I apply your changes but it creates duplicity in all my entities with suffix _2.
On the other hand it works without any problems, also for PG and temperature sensors.

Koky05 commented:
@karelkryda Yes, it was acceptable for me. In my case it was only 45 entities to delete and another 45 entities to configure (area and name change). But I can imagine that can be more complex in some scenarios.

Koky05 commented:
@karelkryda personally I will preffer to leave it as it was to not to break all automations. For example all my thermostats change entity id not only by adding suffix, but also add prefix. So at the end of the day I was rollbacked all your changes in naming convention.

karelkryda commented:
I'll mark this PR as a draft and try to find a way to make changes to unique IDs backwards compatible. Something like when generating entities, I'll try to find existing entities and use the old way of generating unique IDs if they match, otherwise use the new one. That way the old entities should still be functional and the new ones should use the new corrected scheme.

karelkryda commented:
Very good question @Pigotka, I would say yes.
Respectively, it looks like all entities are stored in the registry. It also contains a unique ID. If it changes, it creates new entities using the generator and adds `_2` in case of a conflict.
I'm just looking at the registry code and it's theoretically possible to delete old entities during configuration migration, but I don't think that's entirely appropriate.

I am currently trying to write code that will return a unique ID according to the old schema if it is an existing entity, otherwise it will return the new schema. This should ensure backwards compatibility.

What do you think about this?

Pigotka commented on custom_components/jablotron_cloud/const.py:
```diff
@@ -1,22 +1,26 @@
-
-SERVICES_WITHOUT_PG = ["FUTURA2", "AMBIENTA", "VOLTA", "LOGBOOK"]
-
-class Actions(StrEnum):
-    """Actions to control sections."""
-
-    ARM = "ARM"
-    DISARM = "DISARM"
-    PARTIAL_ARM = "PARTIAL_ARM"
```
Why hardcoding the API commands instead of having constants?

Pigotka commented on custom_components/jablotron_cloud/alarm_control_panel.py:
```diff
@@ -1,208 +1,280 @@
-        self._attr_alarm_state = AlarmControlPanelState.DISARMING
-        self.schedule_update_ha_state()
+            # Set state to disarming if disarm action was successful
+            if disarm_successful:
```
What will HA do in case this is False? Do we need to give some error? Same for all the alarm actions.

Pigotka commented on custom_components/jablotron_cloud/config_flow.py:
```diff
@@ -1,73 +1,131 @@
+            vol.Required(CONF_PASSWORD): str,
+            vol.Optional(CONF_PIN, default=config.get(CONF_PIN, "")): str,
+            vol.Optional(CONF_FORCE_UPDATE, default=config.get(CONF_FORCE_UPDATE, True)): bool,
+            vol.Optional(CONF_SCAN_INTERVAL, default=config.get(CONF_SCAN_INTERVAL, 30)): int,
```
I'm not a fan on allowing user to change scan interval.
1) Minimum is 30 and users will try to add lower values. This should not be allowed. We might overload the API and get banned.
2) Why would you need higher interval? I need lower but higher does nto make any sense.

karelkryda replied:
I think it should be the user's concern what they set. 
Originally it was in the README section of the integration limits. I added it as an option for the user to customize the settings. Maybe someone will be fine with a higher limit, maybe someone will want 20 seconds.

Either way, the 30 second interval is used by default so the user doesn't have to change anything.

Pigotka commented on custom_components/jablotron_cloud/config_flow.py:
```diff
@@ -1,73 +1,131 @@
+            vol.Optional(CONF_PIN, default=config.get(CONF_PIN, "")): str,
+            vol.Optional(CONF_FORCE_UPDATE, default=config.get(CONF_FORCE_UPDATE, True)): bool,
+            vol.Optional(CONF_SCAN_INTERVAL, default=config.get(CONF_SCAN_INTERVAL, 30)): int,
+            vol.Optional(CONF_TIMEOUT, default=config.get(CONF_TIMEOUT, 15)): int
```
Also do not like to have this settings. Why do you think we need it?

1. Tt exists only because jablotronpy was no reliable. It should have been always responsibility of jablotronpy to return reliable data in time or exception.
2. User has no idea why the number should be and what it means for him.
3. I was hopping you get rid of this completely by updating jablotronpy.

karelkryda replied:
This is not a JablotronPy problem, but a Jablotron API problem. Unfortunately, often the request takes a long time or gets stuck in a state where it keeps an open connection to the Cloud.
Even now, unfortunately, this happens and so I think it is useful to have a limit on when the API should get the data. What happens now is that if the limit is reached, at least the data that was able to be retrieved is applied - which is new.

And as I wrote above, the user sees some defaults and if he doesn't understand them, he shouldn't modify them. By default, the data is updated every 30 seconds and has 15 seconds to update.

Pigotka commented:
Regarding the breaking changes due to unique names. I do not mind to for everyone deletes all old entries I have done that many times for different integrations. But my question is why do we need it? Is it a requirement not to have a space? Does it cause any problems right now?

karelkryda commented on custom_components/jablotron_cloud/types.py:
```diff
@@ -0,0 +1,15 @@
+
+from jablotronpy import JablotronSections, JablotronProgrammableGates, JablotronThermoDevice
+
+JablotronServiceData = TypedDict(
```
nope, this is internal type used by integration to store API data - not related to JablotronPy :)

karelkryda commented on custom_components/jablotron_cloud/jablotron.py:
```diff
@@ -0,0 +1,27 @@
+    def get_bridge(self) -> Jablotron:
+        """Return Jablotron bridge instance."""
+
+        bridge = Jablotron(self._username, self._password, self._default_pin)
```
since I removed retries and autologin feature from JablotronPy - it should not be part of it - I needed to handle logins by the integration.
The reason to create new instance every time is, that API cuts sessions older than X. To prevent the issues with expired sessions, I decided to do it this way. Also since every request is separate request it does make sense to be called with new session ID.

my opinion :)

karelkryda commented on custom_components/jablotron_cloud/__init__.py:
```diff
@@ -1,177 +1,230 @@
+                # Check whether service type is supported
+                if service_type in UNSUPPORTED_SERVICES:
+                    _LOGGER.debug("Service '%s' is not supported, ignoring!", service_type)
 
```
sorry, I'm not quiet sure what is the requested change here :/

karelkryda commented on custom_components/jablotron_cloud/binary_sensor.py:
```diff
@@ -1,110 +1,163 @@
+            # Check whether programmable gate is NOT controllable
+            if gate["can-control"]:
+                _LOGGER.debug("Programmable gate '%s' is controllable, ignoring!", gate_name)
 
```
sorry, I'm not quiet sure what is the requested change here :/

karelkryda commented:
@Pigotka, I added comments to your review, and fixed multiline imports to use ().

I also resolved the issue with entity re-creation and added step to migrate ids with spaces to ids with `_`. Removing this as breaking change, as testing on my instance caused no issues :)

karelkryda commented:
@Pigotka,
oh, okay... I didn't know that. If that's true, their API is really bad.
In that case, I'll set it to a minimum of 30 seconds.
